File: mvn-advent-of-code/src/main/java/advent/twenty_fifteen/Day15ScienceForHungryPeople.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.twenty_fifteen.support.Ingredient;
import advent.utilities.FileUtilities;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Day15ScienceForHungryPeople implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    public Day15ScienceForHungryPeople(String filename) {
        List<Ingredient> ingredients = FileUtilities.readLines(filename, Ingredient::PARSE);
        List<List<Integer>> combinations = populate(ingredients.size(), 100);

        Map<List<Integer>, Integer> cookieMakeup = new HashMap<>();
        Map<List<Integer>, Integer> cookieCalorie = new HashMap<>();

        combinations.forEach(combination -> {
            int capacity = 0;
            int durability = 0;
            int flavor = 0;
            int texture = 0;
            int calories = 0;

            for (int i = 0; i != ingredients.size(); i++) {
                int count = combination.get(i);
                Ingredient ingredient = ingredients.get(i);

                capacity += ingredient.getCapacity() * count;
                durability += ingredient.getDurability() * count;
                flavor += ingredient.getFlavor() * count;
                texture += ingredient.getTexture() * count;
                calories += ingredient.getCalories() * count;
            }

            capacity = (capacity > 0) ? capacity : 0;
            durability = (durability > 0) ? durability : 0;
            flavor = (flavor > 0) ? flavor : 0;
            texture = (texture > 0) ? texture : 0;

            cookieMakeup.put(combination, capacity * durability * flavor * texture);
            cookieCalorie.put(combination, calories);
        });

        List<List<Integer>> validCookies = cookieCalorie.entrySet().stream().filter(e -> e.getValue() == 500).map(e -> e.getKey()).collect(Collectors.toList());

        this.part1Answer = cookieMakeup.values().stream().max(Integer::compareTo).get();
        this.part2Answer = validCookies.stream().map(key -> cookieMakeup.get(key)).max(Integer::compareTo).get();
    }

    private List<List<Integer>> populate(int columns, int target) {
        List<List<Integer>> result = new LinkedList<>();

        int[] n = new int[columns];
        n[0] = 0;

        boolean overflow = false;
        while (!overflow) {
            int total = 0;
            for (int c = 0; c != columns; c++) {
                total += n[c];
            }
            if (total == target) {
                List<Integer> found = new LinkedList<>();
                for (int c = 0; c != columns; c++) {
                    found.add(n[c]);
                }
                result.add(found);
            }
            overflow = increment(n, 0, target);
        }
        return result;
    }

    private boolean increment(int[] data, int column, int max) {
        if (column >= data.length) {
            return true;
        }

        data[column]++;
        if (data[column] > max) {
            data[column] = 0;
            return increment(data, column + 1, max);
        }
        return false;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day15ScienceForHungryPeople.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

case class Ingredient(name: String, capacity: Int, durability: Int, flavor: Int, texture: Int, calories: Int)

case class Recipe(ingredient: Ingredient, amount: Int)

class Day15ScienceForHungryPeople(filename: String) extends DailyProblem[Int, Int] {
  def parse(line: String): Ingredient = {
    val pattern = "([A-Za-z]+): capacity ([-]*[0-9]+), durability ([-]*[0-9]+), flavor ([-]*[0-9]+), texture ([-]*[0-9]+), calories ([-]*[0-9]+)".r
    line match {
      case pattern(name, capacity, durability, flavor, texture, calories) => Ingredient(name, capacity.toInt, durability.toInt, flavor.toInt, texture.toInt, calories.toInt)
    }
  }

  private val ingredients = FileUtilities.readFile(filename, parse)

  private def walk(remaining: List[Ingredient], usage: Int, spaceLeft: Int, recipe: List[Recipe], acc: List[List[Recipe]]): List[List[Recipe]] = {
    if (remaining.tail.isEmpty) {
      (Recipe(remaining.head, spaceLeft) :: recipe) :: acc
    } else {
      if (spaceLeft < 0) {
        Nil
      } else {
        walk(remaining.tail, 0, spaceLeft, Recipe(remaining.head, usage) :: recipe, acc) ::: walk(remaining, usage + 1, spaceLeft - 1, recipe, acc)
      }
    }
  }

  private val combinations = walk(ingredients, 0, 100, List(), List())

  private def multiply(recipe: Recipe): (Int, Int, Int, Int) = {
    (recipe.ingredient.capacity * recipe.amount, recipe.ingredient.durability * recipe.amount, recipe.ingredient.flavor * recipe.amount, recipe.ingredient.texture * recipe.amount)
  }

  private def sum(lhs: (Int, Int, Int, Int), rhs: (Int, Int, Int, Int)): (Int, Int, Int, Int) = {
    lhs match {
      case (a, b, c, d) => rhs match {
        case (e, f, g, h) => (a + e, b + f, c + g, h + d)
      }
    }
  }

  private def score(entry: (Int, Int, Int, Int)): Int = {
    entry match {
      case (a, b, c, d) => (if (a > 0) a else 0) * (if (b > 0) b else 0) * (if (c > 0) c else 0) * (if (d > 0) d else 0)
    }
  }

  private def calories(recipe : List[Recipe]) : Int = {
    recipe.foldLeft(0)((a,r) => a + r.amount * r.ingredient.calories)
  }

  override val part1Answer: Int = combinations.map(entry => score(entry.map(recipe => multiply(recipe)).foldLeft((0, 0, 0, 0))((a, e) => sum(a, e)))).max
  override val part2Answer: Int = combinations.filter(entry => calories(entry) == 500).map(entry => score(entry.map(recipe => multiply(recipe)).foldLeft((0, 0, 0, 0))((a, e) => sum(a, e)))).max
}
